src/main/java/picard/sam/HitsForInsert.java

```diff
@@ -194,45 +195,100 @@ public void setPrimaryAlignment(final int primaryAlignmentIndex) {
      * reads may have been filtered out.
      */
     public void coordinateByHitIndex() {
+        // null HI tag sorts after any non-null.
+        final Comparator<SAMRecord> comparator = Comparator.comparing(
+                record -> record.getIntegerAttribute(SAMTag.HI.name()),
+                Comparator.nullsLast(Comparator.naturalOrder())
+        );
         // Sort by HI value, with reads with no HI going at the end.
-        Collections.sort(firstOfPairOrFragment, comparator);
-        Collections.sort(secondOfPair, comparator);
+        firstOfPairOrFragment.sort(comparator);
+        secondOfPair.sort(comparator);
 
         // Insert nulls as necessary in the two lists so that correlated alignments have the same index
         // and uncorrelated alignments have null in the other list at the corresponding index.
         for (int i = 0; i < Math.min(firstOfPairOrFragment.size(), secondOfPair.size()); ++i) {
-            final Integer leftHi = firstOfPairOrFragment.get(i).getIntegerAttribute(SAMTag.HI.name());
-            final Integer rightHi = secondOfPair.get(i).getIntegerAttribute(SAMTag.HI.name());
-            if (leftHi != null) {
-                if (rightHi != null) {
-                    if (leftHi < rightHi) secondOfPair.add(i, null);
-                    else if (rightHi < leftHi) firstOfPairOrFragment.add(i, null);
-                    // else the are correlated
-                }
-            } else if (rightHi != null) {
-                firstOfPairOrFragment.add(i, null);
-            } else {
-                // Both alignments do not have HI, so push down the ones on the right.
-                // Right is arbitrarily picked to push down.
-                secondOfPair.add(i, null);
+            final Integer hiTagValueOfFirst = firstOfPairOrFragment.get(i).getIntegerAttribute(SAMTag.HI.name());
+            final Integer hiTagValueOfSecond = secondOfPair.get(i).getIntegerAttribute(SAMTag.HI.name());
+
+            if (nonNull(hiTagValueOfFirst) && nonNull(hiTagValueOfSecond)) {
+                if (hiTagValueOfFirst < hiTagValueOfSecond) secondOfPair.add(i, null);
```

Comment:
The original code didn't do this, but the new code should use `{` `}` around all if/else body statements, even if they're a single line.

Reply:
Done [here](https://github.com/broadinstitute/picard/pull/932/commits/e418168597725fdc18640ecc687348efcc2ee169).